handler/src/main/java/io/netty/handler/ssl/PemPrivateKey.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import java.security.PrivateKey;
+
+import javax.security.auth.Destroyable;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.base64.Base64;
+import io.netty.util.AbstractReferenceCounted;
+import io.netty.util.CharsetUtil;
+import io.netty.util.IllegalReferenceCountException;
+
+/**
+ * This is a special purpose implementation of a {@link PrivateKey} that can
+ * only be used in conjunction with {@link OpenSslContext} and it cannot be
+ * used to do crypto stuff in Java land.
+ *
+ * All methods other than what's implemented in {@link PemEncoded} and {@link Destroyable}
+ * throw {@link UnsupportedOperationException}s.
+ *
+ * @see PemEncoded
+ * @see OpenSslContext
+ */
+public final class PemPrivateKey extends AbstractReferenceCounted implements PrivateKey, PemEncoded {
+
+    private static final byte[] BEGIN_PRIVATE_KEY = "-----BEGIN PRIVATE KEY-----\n".getBytes(CharsetUtil.US_ASCII);
+    private static final byte[] END_PRIVATE_KEY = "\n-----END PRIVATE KEY-----\n".getBytes(CharsetUtil.US_ASCII);
+
+    /**
+     * Creates a {@link PemPrivateKey} from a {@link PrivateKey}.
+     */
+    static PemPrivateKey toPEM(PrivateKey key) {
+        if (key instanceof PemPrivateKey) {
+            return (PemPrivateKey) key;
+        }
+
+        ByteBuf encoded = Unpooled.wrappedBuffer(key.getEncoded());
+        try {
+            ByteBuf pem = Unpooled.buffer(encoded.readableBytes());
```

Comment:
Pass a `ByteBufAllocator` as a parameter and use it to allocate this buffer. We can use `UnpooledByteBufAllocator.DEFAULT` by default but at least this gives control over allocations done in this method.